### ISSUE TITLE
Trouble importing certificate with yubihsm-shell

### ISSUE THREAD

**Issue by dbrgn** [Yubico/yubihsm-shell#263]
With yubihsm-shell 2.3.0, I'm having some trouble importing a certificate from a file.

When I import an asymmetric key, this seems to work properly:

```
yubihsm> put asymmetric 1 0 "Code Signing Key" 1 sign-pkcs rsa.keyyyy
Unable to read asymmetric key
yubihsm> put asymmetric 1 0 "Code Signing Key" 1 sign-pkcs rsa.key
Stored Asymmetric key 0x09bc
```

When using a non-existing key filename, an error message is returned.

However, when importing an X509 certificate (as an opaque object), this doesn't seem to look up the file:

```
yubihsm> put opaque 1 0 "Code Signing Cert" 1 none opaque-x509-certificate cert.der
Stored Opaque object 0xd194
yubihsm> put opaque 1 0 "Code Signing Cert" 1 none opaque-x509-certificate cert.derrrrrrr
Stored Opaque object 0x3b1c
```

No matter what filename I pass in, an opaque object is always created. This appears to indicate that the file is not actually being read.

Am I doing something wrong, or is there an issue with the `put opaque` command in yubihsm-shell?

**Comment by dbrgn** [comment]
Today I realized that the certificate _was_ properly imported with the first command. So from this behavior, it seems that when a file with the specified filename is found, it is imported. On the other hand, when no file with the specified filename exists in CWD, the raw bytes(?) are used directly. There is no feedback given to the user on what is actually being done, or where the bytes were imported from. That is... not great UX 🙃

**Comment by qpernil** [comment]
The latest release (found here https://developers.yubico.com/yubihsm-shell/Releases/) will fail like this:
put opaque 0 0 "Code Signing Cert" 1 none opaque-x509-certificate cert.derrrrrrr
Couldn't parse DER-encoded certificate

The version you are using doesn't validate that it's a certificate, but the latest one does. The general input handling is also a bit weird in that it treats any string that denotes a valid file by taking the contents of that file for inputs with format 'i' (input data). If it doesn't find a file it simply takes the string as the input as-is, without any warning. I was considering changing that (particularly since there is also a 'file:' prefix you can add, which will fail if the file is not found.) But alas, it would break your usage when you meant to use a file, making it a breaking change.. Hence it has stayed as it is.

**Comment by qpernil** [comment]
[See PR #230](https://github.com/Yubico/yubihsm-shell/pull/230)

**Comment by dbrgn** [comment]
That's great, thank you! Unfortunately the latest yubihsm-sdk (2021.12c) still contains version yubihsm-shell 2.3.0b, that's why the version I used didn't yet validate the certificate. So an updated release might avoid some of these pitfalls 🙂

**Comment by qpernil** [comment]
Quoting an answer to a different issue here:

yubihsm-sdk version 2021.12c contains yubihsm-shell version 2.3.0 in which this issue is present. The issue has been fixed in yubishm-shell version 2.3.1 which was release in February this year and can be found here: https://developers.yubico.com/yubihsm-shell/Releases/

We are currently moving to a module where the different components of the yubihsm-sdk can be released individually when needed to allow for a smoother release process. Full yubihsm-sdk packages will still be made available regularly but minor and patch versions of the individual components may also be released in the meantime, specially when they don't contain security fixes.
